pkg/crdschema/gen.go

```diff
@@ -0,0 +1,377 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package crdschema
+
+import (
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"reflect"
+	"strings"
+
+	"gopkg.in/yaml.v2"
+
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+
+	crdgen "sigs.k8s.io/controller-tools/pkg/crd"
+	crdmarkers "sigs.k8s.io/controller-tools/pkg/crd/markers"
+	crdschemayaml "sigs.k8s.io/controller-tools/pkg/crdschema/yaml"
+	"sigs.k8s.io/controller-tools/pkg/genall"
+	"sigs.k8s.io/controller-tools/pkg/markers"
+)
+
+var (
+	scheme = runtime.NewScheme()
+	codecs = serializer.NewCodecFactory(scheme)
+)
+
+func init() {
+	utilruntime.Must(apiextensionsv1beta1.AddToScheme(scheme))
+}
+
+// Generator is a genall.Generator that generates CRD schemas.
+type Generator struct {
+	// ManifestsPath is a path with CustomResourceDefinition manifest files
+	ManifestsPath string `marker:"manifests,optional"`
+	// If true, no file is written. The Generate func will return an
+	// error if a file on disk differs from the generated manifest.
+	VerifyOnly bool `marker:",optional"`
+}
+
+var _ genall.Generator = &Generator{}
+
+func (Generator) RegisterMarkers(into *markers.Registry) error {
+	return crdmarkers.Register(into)
+}
+
+func (g Generator) Generate(ctx *genall.GenerationContext) (result error) {
+	p := &crdgen.Parser{
+		Collector: ctx.Collector,
+		Checker:   ctx.Checker,
+	}
+	f := crdgen.Flattener{
+		Parser: p,
+	}
+
+	crdgen.AddKnownTypes(p)
+	for _, root := range ctx.Roots {
+		p.NeedPackage(root)
+	}
+
+	metav1Pkg := crdgen.FindMetav1(ctx.Roots)
+	if metav1Pkg == nil {
+		// no objects in the roots, since nothing imported metav1
+		fmt.Printf("No metav1 import found in any of the API packages.\n")
+		return nil
+	}
+
+	// load existing manifests from manifests/ dir
+	existingCRDsByFileNames, err := crdsFromDirectory(ctx, g.ManifestsPath)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("Found %d CRD manifests.\n", len(existingCRDsByFileNames))
+	existingFileNamesByGroupKind := map[schema.GroupKind]string{}
+	for fn, crd := range existingCRDsByFileNames {
+		existingFileNamesByGroupKind[crd.GroupKind] = fn
+	}
+
+	changedFiles := []string{}
+	for _, groupKind := range crdgen.FindKubeKinds(p, metav1Pkg) {
+		fn, found := existingFileNamesByGroupKind[groupKind]
```

Comment:
`fileName`.  don't use single or double-letter variable names unless they're a method receiver or *blindingly* obvious (`fn` could be `fileName`, `function`, etc)

Reply:
`s/found/existsAsFile`.

Reply:
done